pubtools/pulplib/_impl/model/maintenance.py

```diff
@@ -0,0 +1,147 @@
+    started = attr.ib(default=None, type=str)
+    """When did the maintenance start"""
+
+
+@attr.s(kw_only=True)
+class MaintenanceReport(object):
+    """Represent the maintenance status of pulp"""
+
+    _SCHEMA = load_schema("maintenance")
+
+    last_updated = attr.ib(default=_iso_time_now(), type=str)
+    """:class:`~datetime.datetime` in UTC when this report was last updated,
+    if it's the first time the report is created, current time is used."""
+
+    last_updated_by = attr.ib(default="Content Delivery", type=str)
+    """Person/party who updated the report last time"""
+
+    repos = attr.ib(default={}, type=dict)
+    """A dictionary contains repo_id: :class:`MaintainedRepo` pairs, indicates
+    which repositories are in maintenance mode and details.
+    If empty, then it means no repositories is in maintenance mode.
+    """
+
+    @classmethod
+    def from_data(cls, data):
+        """Create a new report with a dictionary"""
+        try:
+            jsonschema.validate(instance=data, schema=cls._SCHEMA)
+        except jsonschema.exceptions.ValidationError as error:
+            LOG.exception("%s.from_data invoked with invalid data", cls.__name__)
+            raise InvalidDataException(str(error))
+
+        maintenance = cls(
+            last_updated=data["last_updated"],
+            last_updated_by=data["last_updated_by"],
+            repos={},
+        )
+        for repo_id, details in data["repos"].items():
+            maintenance.repos[repo_id] = MaintainedRepo(**details)
+
+        return maintenance
+
+    def json(self):
+        """Output a json format report.
+
+        Returns:
+            A string contains the maintenance report in json format.
+        """
+        report = {
+            "last_updated": self.last_updated,
+            "last_updated_by": self.last_updated_by,
+            "repos": {},
+        }
+
+        for repo_id, details in self.repos.items():
+            report["repos"].update(
+                {
+                    repo_id: {
+                        "message": details.message,
+                        "owner": details.owner,
+                        "started": details.started,
+                    }
+                }
+            )
+
+        return json.dumps(report, indent=4, sort_keys=True)
+
+    def add(self, repo_ids, **kwargs):
+        """Add entries to maintenance report and update the timestamp. Every
+        entry added to the report represents a repository in maintenace mode.
+
+        Args:
+            repo_ids (list[str]):
+            A list of repositories ids that will be added to the report.
+
+            Optional keyword args:
+                message (str):
+                    Reason why put the repo to maintenance.
+                owner (str):
+                    Who set the maintenance mode.
+
+        """
+        message = kwargs.get("message") or "Maintenance mode is enabled"
+        owner = kwargs.get("owner") or "Content Delivery"
```

Comment:
I suggest "pubtools-pulplib" is a better name than "Content Delivery".  It more clearly communicates that we only know which tool did the update and we don't know anything else.

However, maybe we can do even better again.  What do you think of instead defaulting to something like $USER@$HOSTNAME?  e.g. "root@rcm-pub-worker-01.example.com" - would give us a better hint at what set a repo in maintenance mode.

In practice, of course we want the callers to provide this argument whenever possible.